src/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/InteropHelpers.cs

```diff
@@ -228,7 +227,47 @@ internal static IntPtr GetDelegateFunctionPointer()
             return fnPtr;
         }
 
-        internal static IntPtr GetStubForPInvokeDelegate(Delegate del)
+        /// <summary>
+        /// Retrieves the current delegate that is being called
+        /// </summary>
+        public static T GetCurrentCalleeDelegate<T>() where T : class // constraint can't be System.Delegate
+        {
+            //
+            // RH keeps track of the current thunk that is being called through a secret argument / thread
+            // statics. No matter how that's implemented, we get the current thunk which we can use for
+            // look up later
+            //
+            IntPtr pContext = RuntimeImports.GetCurrentInteropThunkContext();
+
+            Debug.Assert(pContext != null);
+
+            GCHandle handle;
+            unsafe
+            {
+                // Pull out Handle from context
+                handle = (*((ThunkContextData*)pContext)).Handle;
+
+            }
+
+            T target = InteropExtensions.UncheckedCast<T>(handle.Target);
+
+            //
+            // The delegate might already been garbage collected
+            // User should use GC.KeepAlive or whatever ways necessary to keep the delegate alive
+            // until they are done with the native function pointer
+            //
+            if (target == null)
+            {
+                Environment.FailFast(
+                    "The corresponding delegate has been garbage collected. " +
```

Comment:
If this can be caused by user error, I think localizable message is preferable. Is that state at this point really that bad that we can't have a localizable error message?